### DISCUSSION TITLE
Proportionally scale arbitrary canvas to a fixed width

### DISCUSSION THREAD

**Discussion by noahjutz** [cetz-package/cetz#689] [Q&A] [answered]
I'm looking for a canvas-agnostic solution for scaling any canvas to a fixed width (e. g. the page width), such that the leftmost visible element borders to the left edge of the canvas and rightmost element borders to the right edge of the canvas. The canvas should vertically scale proportionally, so the aspect ratio is maintained.

Or, more simply put, I want to resize my graphs like you would in any image editing software:

[Cropping and scaling an image in inkscape](https://github.com/user-attachments/assets/a9fe14af-d8fc-4da6-b1eb-1fbdb692e594)

Canvases can be scaled to the parent's size by using a relative unit as its `length` (e. g. 10%), but I would like a solution that works independently of the canvas' content bounds; using `length: 10%`, for example, would impose the limitation of having a fixed 10 units horizontally available.

The `set-viewport` transformation function allows for mapping arbitrary bounds to a fixed width, but this requires manually passing the bounds of the content - a task that could be automated. To complicate matters further, you'd have to measure variably sized shapes like `content` somehow.

![measuring text in CeTZ](https://github.com/user-attachments/assets/22efefdb-bb1d-4c75-8b05-2fd21f0ffa8d)

One potential solution might be to resolve the position of each element, caching the min and max x and y values, and somehow retroactively applying a transformation based on those values. But maybe there's a much more simple solution I'm missing.

**Comment by johannes-wolf** [comment, marked as answer]
Something like:
```typ
#import "@preview/cetz:0.2.2"
#set page(width: 12cm, height: 10cm)

#let scale-to(body, target-width: 1) = {
  (ctx => {
    import cetz.draw: *

    let (ctx: _, drawables, bounds) = cetz.process.many(ctx, cetz.util.resolve-body(ctx, body))
    
    let bounds = cetz.vector.sub(bounds.high, bounds.low)
    let ratio = bounds.at(1) / bounds.at(0)

    return cetz.process.many(ctx, {
      group({
        set-viewport((0, 0, 0), (target-width, target-width * ratio, 1), bounds: bounds,)
        body
      })
    })
  },)
}

#place(rect(width: 10cm, height: 8cm))

#cetz.canvas(length: 10cm, {
  import cetz.draw: *

  scale-to({
    circle((2,1), fill: red)
    rotate(300deg)
    rect((0,0), (rel: (2, 2)), fill: blue)
  })
})
```

Gives:
![grafik](https://github.com/user-attachments/assets/5e0cd539-6401-4909-be99-1243495f9c69)


But as you have mentioned, finding the right scaling with unscaled content in the canvas is non-trivial.

**Comment by johannes-wolf** [comment]
You could also apply scaling to the content by modifying the `.body` field of drawable with `.type == "content"`.

**Comment by johannes-wolf** [comment]
Alternative: Use Typst' naitve `scale()` function:
```typ
#place(rect(width: 10cm, height: 8cm))

#let c = cetz.canvas({
  import cetz.draw: *

  circle((1,1), fill: red)
  content((5,0), [Some test text is here. \ What do you think?])
  rotate(280deg)
  rect((1,3), (rel: (2, 2)), fill: blue)
})

#place(context {
  let (width, height) = measure(c)
  scale(10cm / width * 100%, place(c))
})
```

![grafik](https://github.com/user-attachments/assets/d0835e92-3b18-4d5b-940a-546292a79204)


The downside is, that it scales the stroke-width of your canvas.

**Reply by NicoWeio** [reply]
As a drop-in solution, this is what I ended up with to fine-adjust figures to text width. The 100% could of course be made an argument, if needed.
```typst
#let scale-to-width(content) = context {
  let target = page.width - page.margin.left.length - page.margin.right.length
  let (width, height) = measure(content)
  scale(target / width * 100%, reflow: true, content)
}
```

**Reply by rajayonin** [reply]
> As a drop-in solution, this is what I ended up with to fine-adjust figures to text width. The 100% could of course be made an argument, if needed.
> 
> ```typst
> #let scale-to-width(content) = context {
>   let target = page.width - page.margin.left.length - page.margin.right.length
>   let (width, height) = measure(content)
>   scale(target / width * 100%, reflow: true, content)
> }
> ```

This only works if you setup the page with `#set page(margin: (left: ..., right: ...))`.

If you want a generic solution:


```typst
/// Scales content to the desired width/height relative to its container.
///
/// - body (content): The content to scale
/// - width (ratio, auto): Desired width (relative to its container)
/// - height (ratio, auto): Desired heigth (relative to its container)
/// -> content
#let scale-to-container(body, width: auto, height: auto) = layout(ly => {
  let body-size = measure(body)

  scale(
    x: if width == auto { auto } else { ly.width / body-size.width * width },
    y: if height == auto { auto } else { ly.height / body-size.height * height },
    reflow: true,
    body,
  )
})
```

**Edit:** I was [made aware that `layout` exists](https://forum.typst.app/t/how-to-compute-the-current-pages-textwidth/6523/2?u=rajayonin), so I updated my solution.
<details>
<summary>Previous solution</summary>

```typst
/// Computes the current page's textwidth
///
/// - page (function): Typst's `page` function
/// -> length
#let compute-textwidth(page) = {
  // Margins are set automatically to 2.5/21 times the smaller dimension of the
  // page.
  let default-margin = (2.5 / 21) * calc.min(page.width, page.height)

  if page.margin == auto {
    return page.width - (2 * default-margin)
  }

  if "x" in page.margin {
    return page.width - (2 * page.margin.x.length)
  }

  let left-margin = if "left" in page.margin {
    page.margin.left.length
  } else if "inside" in page.margin {
    page.margin.inside.length
  } else if "rest" in page.margin {
    page.margin.rest.length
  } else { default-margin }

  let right-margin = if "right" in page.margin {
    page.margin.right.length
  } else if "outside" in page.margin {
    page.margin.outside.length
  } else if "rest" in page.margin {
    page.margin.rest.length
  } else { default-margin }

  return page.width - left-margin - right-margin
}

/// Scales content to the desired width.
///
/// - body (content):
/// - width (fraction): desired width
/// -> content
#let scale-to-width(body, width: 100%) = context {
  let body-width = measure(body).width
  scale(compute-textwidth(page) / body-width * width, reflow: true, body)
}
```

</details>
